File: 55_binary_tree_right_side_view.js
/*
Given the root of a binary tree, imagine yourself standing on the right 
side of it, return the values of the nodes you can see ordered from top 
to bottom.
*/

// https://leetcode.com/problems/binary-tree-right-side-view/description/

// find the rightmost node at each level
// level order traversal
var rightSideView = function (root) {
  let res = [];
  var levelOrder = function (node) {
    if (!node) return;
    let queue = [node];
    while (queue.length) {
      let len = queue.length; // number of nodes at this level
      for (let i = 0; i < len; i++) {
        let curr = queue.shift();
        if (i === len - 1) res.push(curr.val);
        if (curr.left) queue.push(curr.left);
        if (curr.right) queue.push(curr.right);
      }
    }
  };
  levelOrder(root);
  return res;
};

File: 55_binary_tree_right_side_view.py
# Given the root of a binary tree, imagine yourself standing on the right 
# side of it, return the values of the nodes you can see ordered from top 
# to bottom.

# https://leetcode.com/problems/binary-tree-right-side-view/

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        import collections
        if not root: return []
        ans, queue = [], collections.deque([root])
        while queue:
            size = len(queue)
            for i in range(size):
                cur = queue.popleft()
                if i == size - 1:
                    ans.append(cur.val)
                if cur.left: queue.append(cur.left)
                if cur.right: queue.append(cur.right)
        return ans